sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala

```diff
@@ -0,0 +1,413 @@
+            val mergedPlan = Project(mergedProjectList, mergedChild)
+            mergedPlan -> newOutputMap
+          }
+        case (np, cp: Project) =>
+          tryMergePlans(np, cp.child).map { case (mergedChild, outputMap) =>
+            val (mergedProjectList, newOutputMap) =
+              mergeNamedExpressions(np.output, outputMap, cp.projectList)
+            val mergedPlan = Project(mergedProjectList, mergedChild)
+            mergedPlan -> newOutputMap
+          }
+        case (np: Project, cp) =>
+          tryMergePlans(np.child, cp).map { case (mergedChild, outputMap) =>
+            val (mergedProjectList, newOutputMap) =
+              mergeNamedExpressions(np.projectList, outputMap, cp.output)
+            val mergedPlan = Project(mergedProjectList, mergedChild)
+            mergedPlan -> newOutputMap
+          }
+        case (np: Aggregate, cp: Aggregate) if supportedAggregateMerge(np, cp) =>
+          tryMergePlans(np.child, cp.child).flatMap { case (mergedChild, outputMap) =>
+            val mappedNewGroupingExpression =
+              np.groupingExpressions.map(mapAttributes(_, outputMap))
+            if (ExpressionSet(mappedNewGroupingExpression) ==
+              ExpressionSet(cp.groupingExpressions)) {
+              val (mergedAggregateExpressions, newOutputMap) =
+                mergeNamedExpressions(np.aggregateExpressions, outputMap, cp.aggregateExpressions)
+              val mergedPlan =
+                Aggregate(cp.groupingExpressions, mergedAggregateExpressions, mergedChild)
+              Some(mergedPlan -> newOutputMap)
+            } else {
+              None
+            }
+          }
+
+        // Merging general nodes is complicated and this implementation supports only those nodes in
+        // which the order and the number of output attributes are not relevant (see
+        // `supportedMerge()` whitelist).
+        // Also, this implementation supports only those nodes in which children can be merged in
+        // the same order.
+        case (np, cp) if supportedMerge(np) && np.getClass == cp.getClass &&
+          np.children.size == cp.children.size =>
+          val merged = np.children.zip(cp.children).map {
+            case (npChild, cpChild) => tryMergePlans(npChild, cpChild)
+          }
+          if (merged.forall(_.isDefined)) {
+            val (mergedChildren, outputMaps) = merged.map(_.get).unzip
+            val outputMap = AttributeMap(outputMaps.map(_.iterator).reduce(_ ++ _).toSeq)
+            val mappedNewPlan = mapAttributes(np.withNewChildren(mergedChildren), outputMap)
+            val mergedPlan = cp.withNewChildren(mergedChildren)
+            if (mappedNewPlan.canonicalized == mergedPlan.canonicalized) {
+              Some(mergedPlan -> outputMap)
+            } else {
+              None
+            }
+          } else {
+            None
+          }
+
+        // As a follow-up, it would be possible to merge `CommonScalarSubqueries` nodes, which would
```

Comment:
perhaps I am not catch all. wouldn't it be straight forward if you put the `Common` under the root node and merge reference plans there ?

Reply:
I don't think it is straightforward to refer to a subquery in a common root node of the whole plan, from another subquery, but it is probably doable. Unfortunately, even this simpler change hasn't got much reviews...

Anyway, I will update this PR soon to resolve the conflicts.

Reply:
I guess correlated queries worths another PR ( which I guess we can consider queries de-correlated first). but here since you are targeting scalar sub queries, why not just gather these merged scalar queries and compute them first and then reference them back in the main query?  guess it is essentially the same with current approach

Reply:
guess it involves more execution work though

Reply:
> I guess correlated queries worths another PR

That's what I would like to do next if this PR is accepted.

Reply:
OK, sounds good